### ISSUE TITLE
Improving buffering of SyncedCollections

### ISSUE THREAD

**Issue by vyasr** [glotzerlab/signac#454]
I've been aware of this issue for a while, but I haven't had a chance to benchmark it until working on #453, so I'm just getting around to documenting. While the examples below are for the JSONDict class in master, I only plan to implement any future changes in the new SyncedCollection classes since those are nearing completion.

Currently, while our buffering strategy substantially improves performance on slow filesystems, it isn't particularly useful on fast filesystems (see #453). While it's not surprising that the filesystem speed plays a big role, the fact that buffering can actually slow down our code when used on a fast filesystem shows how much overhead our method incurs. Benchmarking the buffered behavior on a fast filesystem reveals that the bulk of the time is spent on JSON serialization and deserialization (which has to happen on every operation) as well as updating the in-memory representation to match, which has to do a recursive check of the entire dictionary's contents. These operations are extremely expensive. We can easily test this with the current `JSONDict` using the instance-level buffering as currently implemented, which basically just disables all synchronization processes. Consider the following two snippets:

```
d = JSONDict('foo.json')

# Option 1
with buffer_reads_writes():
    for i in range(1000):
        d[str(i)] = i

# Option 2
with d.buffered() as db:
    for i in range(1000):
        db[str(i)] = i
```

On my machine, **Option 2 is ~100x faster**. The reason is that while both of these involve only 1 write to disk, option 1 also has to update the cached data every time, then reload it prior to the call to `d._data.__setitem__`. That indicates to me that we need to seriously reconsider our buffering strategy. 

The basic buffering strategy employed by the per-instance buffering has two problems:

1. It offers no protection against external modifications of the underlying file, nor does it have any way of synchronizing between multiple instances of the file.
2. It has no concept of maximum buffer size. This is fine for a single object, since no matter how big it is we have to write it out at some point and JSON has no concept of incremental changes and will dump the entire file at once. However, this may not be acceptable if we're opening a large number of files and modifying all of them; in addition to the large amount of actual file modification, there's also the overhead of opening and closing many file handles, and some file systems might choke on huge number of writes.

One possible solution is to observe that in many use cases, client code (e.g. signac-flow) may be able to guarantee (or at least assume that the user is staying within safe usage) that (1) is not an issue. In that case, we only have to consider (2). In that case, we might be able to get away with "buffering" by just making in-memory modifications to collections without writing to any centralized buffer at all. In that case, we could use a simpler metric of total process resource usage such as the number of distinct collections that have been changed during buffering up to that point and force-writing all of them when the number gets too large. This feature would be relatively easy to add to the new buffering framework for SyncedCollections.

A much smaller change would be to replace JSON encoding with pickling, which should be safe within the context of buffering (no security issues since it's purely internal/in-memory with no external files being read) and substantially faster. I'm open to other solutions, but given the massive performance implications I think that we need to consider employing some alternative buffering protocol.

**Comment by bdice** [comment]
**Buffering strategy:** I agree that your benchmark of executing many small edits to a single document should be faster using the document-specific buffer. However, in my understanding, the use case for buffering in signac-flow is iterating over thousands of state points in multiple loops, and attempting to avoid re-loading. I might see the benefit if we used this approach for a project document with many status updates, but probably not at the job-level.

**Buffering serialization:** I agree it would be nice if we used another approach instead of recursive conversions with JSON serialization/deserialization. I think that's worth testing, regardless of whether we use global or file-local buffering. The only catch is that (depending on implementation) we might not be able to detect whether the contents are _possible to JSON encode_ until the point where the buffer writes. That would be a lazy validation approach and might still be valid. It may also be possible to ensure JSON-encodability only on the _delta_ of data being updated in the buffer.

**Comment by vyasr** [comment]
On the branch `feature/synced_collections/memory_buffering` I've implemented a relatively simple solution that actually avoids some limitations that I previously foresaw. It doesn't have issue (1) and can keep multiple instances pointing to the same file in sync. It doesn't currently have any concept of a max buffer size, but it would be quite simple to implement something that _approximates_ this using the size of the file on disk when first loaded to the buffer at basically no additional performance cost. While this approximation would be quite poor if we started with an empty file, it would be easy to provide a simple API for calling code to provide some sort of "growth factor" to indicate how big objects can be expected to get. While this isn't ideal, I'd say the performance benefits might pay off depending on the use case.

The basic idea with this buffering is direct memory-sharing: objects directly inserting their local base collections (`self._data`) into the buffer, and if another object tries to load that same file then it repoints its local data to the same underlying dict in the buffer. No serialization to/from JSON occurs, and it also bypasses expensive calls to `_update` every time data is loaded from the buffer. 

You're right that I probably chose an unfair access pattern. Here are new benchmarks that attempt to address your concern about the access pattern. I test for different number of reads/writes on different number of dictionaries, keeping the total number of operations constant. I test on three different file systems: my Mac SSD, our work NFS system mounted on the vislab, and the scratch filesystems also mounted on the Vislab. I was pretty quick and dirty with the implementation, so there are some optimizations to be had. In particular, I'm definitely doing twice as many os.stat calls when initially loading objects, so for large numbers of files my new version is much less efficient than it could be, particularly on NFS where metadata access is very slow (hence the benchmarks on NFS below where the old method wins). My benchmarks also aren't terribly rigorous since I only do 10 loops, so they can vary a decent bit if I rerun, but the results don't change the takeaways meaningfully.

| System             |  Operation | Number of dicts | Elements per dict  | Speed multiplier |
|------------------|------|---------------|-------------------|---------------------------------|
| Mac SSD  | setitem   | 1 | 1000 | 35.28 |
| Mac SSD  | setitem   | 10 | 100 | 30.67 |
| Mac SSD | setitem   | 100 | 10 | 12.88 |
| Mac SSD | setitem   | 1000 | 1 | 2.86 |
| Mac SSD | getitem   | 1 | 1000 | 142.22 |
| Mac SSD  | getitem   | 10 | 100 | 67.93 |
| Mac SSD | getitem   | 100 | 10 | 9.98 |
| Mac SSD | getitem   | 1000 | 1 | 2.33 |
| Linux NFS  | setitem   | 1 | 1000 | 18.27 |
| Linux NFS  | setitem   | 10 | 100 | 7.25 |
| Linux NFS  | setitem   | 100 | 10 | 1.32 |
| Linux NFS  | setitem   | 1000 | 1 | 0.15 |
| Linux NFS  | getitem   | 1 | 1000 | 50.06 |
| Linux NFS  | getitem   | 10 | 100 | 7.67 |
| Linux NFS  | getitem   | 100 | 10 | 0.93 |
| Linux NFS  | getitem   | 1000 | 1 | 0.16 |
| Linux scratch  | setitem   | 1 | 1000 | 22.24 |
| Linux scratch  | setitem   | 10 | 100 | 21.52 |
| Linux scratch  | setitem   | 100 | 10 | 17.35 |
| Linux scratch  | setitem   | 1000 | 1 | 6.60 |
| Linux scratch  | getitem   | 1 | 1000 | 95.52 |
| Linux scratch  | getitem   | 10 | 100 | 76.02 |
| Linux scratch  | getitem   | 100 | 10 | 22.16 |
| Linux scratch  | getitem   | 1000 | 1 | 6.58 |


Regarding the validation, I think I'm OK with being lazy; the file on disk will by definition still be valid.

**Comment by vyasr** [comment]
Closed in #462 .
